### PR TITLE
swarm: bzz-list, bzz-raw and bzz-immutable schemes (#150)

### LINKED ISSUE #150: List requests should use url scheme instead of query parameter
For the list request type we should use the `bzzl://` url scheme instead of the `list=true` query parameter.

## Reason
We decided url schemes are the way to go, see: https://gist.github.com/gbalint/ce09ada0bebc38f3a8a3dd2ce6c6299f

## Question
Should we keep backward compatibility with the query parameter (at least for a while)?

### PR DESCRIPTION
**Pull request by janos**
Replace query parameter list=true for listing all files contained
in a swarm manifest with a new URL scheme bzz-list. This change
does not provide backward compatibility.

Replace bzzr and bzzi schemes with bzz-raw and bzz-immutable.
New URI shemes are added and old ones are deprecated, but not removed.
Old schemes bzzr and bzzi are functional for backward compatibility.

Related gist: https://gist.github.com/gbalint/ce09ada0bebc38f3a8a3dd2ce6c6299f.

This PR is closely related to the https://github.com/ethersphere/go-ethereum/pull/163 PR.

Closes https://github.com/ethersphere/go-ethereum/issues/150 issue.

### CODE REVIEW

**Comment by cobordism** [comment]
I was thinking, it is common in gnu programs to have short and long cli flags. For example we might have `-x` and `--extract`.

Could the same be true here, where `bzzx` is synonymous with `bzz-extract`?

**Comment by gbalint** [comment]
Ok, so let's agree on the functionality, poor @janos goes back and forth with the implementation....
So let's keep the short and long ones too?

**Comment by janos** [comment]
I am not absolutely sure that we need bot short and long ones. I see pros to have them, but can think of at least two cons:
  - user experience - it may be confusing for users
  - http redirects - we would need to make additional effort to preserve the request's scheme in redirect.

Just a few thoughts. Otherwise, I do not object the idea. Schemes bzzr and bzzi will be kept, and new bzzl will be added. Also, I will add bzzh to https://github.com/ethersphere/go-ethereum/pull/163, as well. If you agree?

**Comment by cobordism** [comment]
For the record, I prefer the longer format.
(My last comment was made in the interests of backward compatibility)

I have no strong opinions on this though except that bzzi and bzzl look too similar and that short names are not descriptive.

**Comment by janos** [comment]
Backward compatibility for bzzr and bzzi schemes are not a problem. Reverting the last commit will bring them back. I am fine with that, too. If you agree?

**Comment by cobordism** [comment]
I think currently we are in the mood of breaking backward compatibility for the next release anyway so...

Here's how I feel:
If we want to merge it to upstream master now, we should keep the short names,
If we hold off merging it until we merge all the other breaking changes for our 0.3 release (new hasher, new syncer etc.) then by all means, remove the short version.

The problem is that we don't have a difference between "last release" and "nightly" for swarm. Maybe that's something we should think about after 0.3

> If you agree?

I don't even feel authorised to make this decision :)

**Comment by janos** [comment]
I think that removing support for short names will be trivial for 0.3 (breaking) release. So, we can keep them until then.

**Review by cobordism — APPROVED** [review]

**Comment by gbalint** [comment]
Please resubmit it to https://github.com/ethereum/go-ethereum then
👍

**Review by lmars — APPROVED** [review]

**Comment by gbalint** [comment]
Merged to https://github.com/ethereum/go-ethereum
PR: https://github.com/ethereum/go-ethereum/issues/15667

**Pull request closed without merging**
